masq_lib/src/percentage.rs

```diff
@@ -1,77 +1,226 @@
+    where
+        N: PercentageInteger,
+        <N as TryFrom<i8>>::Error: Debug,
+        N: TryFrom<u32>,
+        <N as TryFrom<u32>>::Error: Debug,
+        i16: TryFrom<N>,
+        <i16 as TryFrom<N>>::Error: Debug,
+    {
+        let multiples = match N::try_from(self.multiples_of_100_percent) {
+            Ok(num) => num,
+            Err(_) => return Err(BaseTypeOverflow {}),
+        };
+
+        let by_wholes = match num.checked_mul(&multiples) {
+            Some(num) => num,
+            None => return Err(BaseTypeOverflow {}),
+        };
+
+        let by_remainder = self.degrees_from_remainder.of(num);
+
+        match by_wholes.checked_add(&by_remainder) {
+            Some(res) => Ok(res),
+            None => Err(BaseTypeOverflow {}),
+        }
+    }
 }
 
-impl Percentage {
-    pub fn new(num: u8) -> Self {
-        match num {
-            0..=100 => Self { per_cent: num },
-            x => panic!("Accepts only range from 0 to 100 but {} was supplied", x),
+#[derive(Debug, PartialEq, Eq)]
+pub struct BaseTypeOverflow {}
+
+impl TryFrom<u8> for PurePercentage {
+    type Error = String;
+
+    fn try_from(degree: u8) -> Result<Self, Self::Error> {
+        match degree {
+            0..=100 => Ok(Self { degree }),
+            x => Err(format!(
+                "Accepts only range from 0 to 100 but {} was supplied",
+                x
+            )),
         }
     }
+}
 
+impl PurePercentage {
     pub fn of<N>(&self, num: N) -> N
     where
-        N: From<u8> + CheckedMul + CheckedAdd + CheckedDiv + PartialOrd + Integer + Debug + Copy,
+        N: PercentageInteger,
+        <N as TryFrom<i8>>::Error: Debug,
+        i16: TryFrom<N>,
+        <i16 as TryFrom<N>>::Error: Debug,
     {
-        let zero = N::from(0);
-        if num == zero || N::from(self.per_cent) == zero {
+        if let Some(zero) = self.return_zero(num) {
             return zero;
         }
 
-        let a = match N::from(self.per_cent).checked_mul(&num) {
+        let product_before_final_div = match N::try_from(self.degree as i8)
+            .expect("Each type has 100")
+            .checked_mul(&num)
+        {
             Some(num) => num,
             None => return self.handle_upper_overflow(num),
         };
 
-        if a < N::from(10) {
-            return N::from(0);
-        }
+        Self::div_by_100_and_round(product_before_final_div)
+    }
 
-        let rounding = if Percentage::should_be_rounded_down(a) {
-            N::from(0)
+    fn return_zero<N>(&self, num: N) -> Option<N>
+    where
+        N: PercentageInteger,
+        <N as TryFrom<i8>>::Error: Debug,
+    {
+        let zero = N::try_from(0).expect("Each type has 0");
+        if num == zero || N::try_from(self.degree as i8).expect("Each type has 100") == zero {
+            Some(zero)
         } else {
-            N::from(1)
-        };
+            None
+        }
+    }
+
+    fn div_by_100_and_round<N>(num: N) -> N
```

Comment:
If you'd prefer to write something for this fn, then maybe a comment explaining the 'why' behind the rounding will actually help in understanding it better.